openshift-tools/pkg/discover-etcd-initial-cluster/initial-cluster.go

```diff
@@ -105,110 +113,136 @@ func (o *DiscoverEtcdInitialClusterOptions) Validate() error {
 		return err
 	}
 	defer client.Close()
 
-	var targetMember *etcdserverpb.Member
-	var allMembers []*etcdserverpb.Member
+	target := url.URL{
+		Scheme: o.TargetPeerURLScheme,
+		Host:   fmt.Sprintf("%s:%s", o.TargetPeerURLHost, o.TargetPeerURLPort),
+	}
+
 	for i := 0; i < 10; i++ {
 		fmt.Fprintf(os.Stderr, "#### attempt %d\n", i)
-		targetMember, allMembers, err = o.checkForTarget(client)
 
-		for _, member := range allMembers {
-			fmt.Fprintf(os.Stderr, "      member=%v\n", stringifyMember(member))
+		// check member list on each iteration for changes
+		cluster, err := client.MemberList(context.TODO())
+		if err != nil {
+			fmt.Fprintf(os.Stderr,"member list request failed: %v", err)
+			continue
+		}
+		logCurrentMembership(cluster.Members)
+		targetMember, found := checkTargetMember(target,cluster.Members)
+
+		// Condition: unstarted member found
+		// This member is part of the cluster but has not yet started. We know this because the name is populated at
+		// runtime which this member does not have.
+		// Result: provide initial cluster value so etcd can communicate with peers during startup.
+		if found && targetMember.Name == "" {
+			fmt.Print(getInitialCluster(o.TargetName, targetMember, cluster.Members))
+			return nil
 		}
-		fmt.Fprintf(os.Stderr, "      target=%v, err=%v\n", stringifyMember(targetMember), err)
 
-		// we're done because we found what we want.
-		if targetMember != nil && err == nil {
-			break
+		// Condition: member found and has started in the past
+		// Result: empty initial config
+		if found && dataDirExists{
+			return nil
 		}
 
-		fmt.Fprintf(os.Stderr, "#### sleeping...\n")
-		time.Sleep(1 * time.Second)
-	}
+		// Condition: member found and has started in the past but not has no dataDir
+		// A member is not actually gone forever unless it is removed from MemberList or its dataDir is destroyed. Since
+		// this is the latter lets not let etcd start and report the condition as an error.
+		// Result: error
+		if found && !dataDirExists{
+			return fmt.Errorf("member %q dataDir has been destoyed, the member must be removed from the cluster", target.String())
+		}
 
-	switch {
-	case targetMember == nil && memberDirExists:
-		// we weren't able to locate other members and need to return based previous memberDir so we can restart.  This is again the off and on flow.
-		fmt.Fprintf(os.Stderr, "Couldn't get targetMember, but memberDir %s is present on %s. Returning with no error.\n", memberDir, o.TargetName)
-		fmt.Printf(o.TargetName)
-		return nil
+		// Condition: target not found in member list
+		// Return without initial cluster value and start existing member, empty initial config is appropriate here.
+		// possible reasons for this condition.
+		// 1.) member has been removed: in this case etcd will crashloop notifying admin/operator cleanup is required.
+		// 2.) on and off flow
+		// Result: empty initial config
+		if !found && dataDirExists {
+			archiveDataDir(o.DataDir)
+			return nil
+		}
```

Comment:
1. If it is `on and off flow` found will be true (if getClient succeeds, other members are up) and we don't come here.
2. If member is removed, we must return err on L202.

Reply:
yeah we talked about this one I think this needs to be aligned with above